File: submissions/0127-word-ladder/solution.ts
function ladderLength(beginWord: string, endWord: string, wordList: string[]): number {
  if (
    !wordList.includes(endWord) ||
    !endWord ||
    !beginWord ||
    wordList.length === 0
  ) {
    return 0;
  }

  const L = beginWord.length;

  const allComboDict: Record<string, string[]> = {};
  wordList.forEach((word) => {
    for (let i = 0; i < L; i++) {
      const genericWord = word.slice(0, i) + '*' + word.slice(i + 1);
      if (allComboDict[genericWord]) {
        allComboDict[genericWord].push(word);
      } else {
        allComboDict[genericWord] = [word];
      }
    }
  });

  const queue: [string, number][] = [[beginWord, 1]];
  const visited: Record<string, boolean> = {};
  visited[beginWord] = true;

  while (queue.length > 0) {
    const [currentWord, level] = queue.shift()!;

    for (let i = 0; i < L; i++) {
      const intermediateWord = currentWord.slice(0, i) + '*' + currentWord.slice(i + 1);

      for (const word of allComboDict[intermediateWord] || []) {
        if (word === endWord) {
          return level + 1;
        }

        if (!visited[word]) {
          visited[word] = true;
          queue.push([word, level + 1]);
        }

        allComboDict[intermediateWord] = [];
      }
    }
  }

  return 0;
}

File: submissions/0127-word-ladder/solution.py
from collections import defaultdict, deque
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0

        def differs_by_one(str1, str2):
            differs = 0

            for i in range(len(str1)):
                if str1[i] != str2[i]:
                    differs += 1
            
            return differs == 1
           
        wordSet = set(wordList)
        queue = deque([beginWord])
        seen = set()
        ans = 1

        while queue:
            for _ in range(len(queue)):
                node = queue.popleft()

                if node == endWord:
                    return ans
            
                for word in wordSet:
                    if word not in seen and differs_by_one(word, node):
                        seen.add(word)
                        queue.append(word)
                    
            ans += 1

        return 0
